File: java/chaincue-real-estate/src/main/java/com/chaincuerealestate/chaincuerealestate/services/DTOBuilderHelpers/HouseHelper.java
package com.chaincuerealestate.chaincuerealestate.services.DTOBuilderHelpers;

import com.chaincuerealestate.chaincuerealestate.domains.House;
import com.chaincuerealestate.chaincuerealestate.services.HouseService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Function;

@RequiredArgsConstructor
@Component
public class HouseHelper {
    private final HouseService houseService;

    public <B> Function<B, B> updateDTOBuilderWithHouseByHouse(Function<B, House> getHouse, BiConsumer<B, House> setHouse) {
        return dtoBuilder -> {
            House reservations = houseService.findById(getHouse.apply(dtoBuilder).getId());
            setHouse.accept(dtoBuilder, reservations);
            return dtoBuilder;
        };
    }

    public <B> Function<B, B> updateDTOBuilderWithHouses(BiConsumer<B, List<House>> setHouses) {
        return dtoBuilder -> {
            List<House> houses = houseService.findAll();
            setHouses.accept(dtoBuilder, houses);
            return dtoBuilder;
        };
    }
}

File: kotlin/chaincue-real-estate/src/main/kotlin/com/example/chaincuerealestate/services/DTOBuilderHelpers/HouseHelper.kt
package com.example.chaincuerealestate.services.DTOBuilderHelpers

import com.example.chaincuerealestate.domains.House
import com.example.chaincuerealestate.services.HouseService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class HouseHelper(@Autowired private val houseService: HouseService) {
    fun <B> updateDTOBuilderWithHouses(setHouses: (B, List<House>) -> Unit): (B) -> B {
        return { dtoBuilder ->
            val houses = houseService.findAll()
            setHouses(dtoBuilder, houses)
            dtoBuilder
        }
    }

    fun <B> updateDTOBuilderWithHouseByHouseId(houseId: String, setHouse: (B, House) -> Unit): (B) -> B {
        return { dtoBuilder ->
            val house = houseService.findById(houseId)
            setHouse(dtoBuilder, house)
            dtoBuilder
        }
    }

    fun <B> updateDTOBuilderWithHouseByHouse(getHouse: (B) -> House, setHouse: (B, House) -> Unit): (B) -> B {
        return { dtoBuilder ->
            val house = houseService.findById(getHouse(dtoBuilder).id)
            setHouse(dtoBuilder, house)
            dtoBuilder
        }
    }

}

